stdlib/in_channel.mli

```diff
@@ -0,0 +1,144 @@
+  | Open_text        (** open in text mode (may perform conversions). *)
+  | Open_nonblock    (** open in non-blocking mode. *)
+(** Opening modes for {!open_gen}. *)
+
+val stdin : t
+(** The standard input for the process. *)
+
+val open' : string -> t
+(** Open the named file for reading, and return a new input channel on that
+    file, positioned at the beginning of the file. *)
+
+val open_bin : string -> t
+(** Same as {!open'}, but the file is opened in binary mode, so that no
+    translation takes place during reads. On operating systems that do not
+    distinguish between text mode and binary mode, this function behaves like
+    {!open'}. *)
+
+val open_gen : open_flag list -> int -> string -> t
+(** [open_gen mode perm filename] opens the named file for reading, as described
+    above. The extra arguments [mode] and [perm] specify the opening mode and
+    file permissions.  {!open'} and {!open_bin} are special cases of this
+    function. *)
+
+val seek : t -> int64 -> unit
+(** [seek chan pos] sets the current reading position to [pos] for channel
+    [chan]. This works only for regular files. On files of other kinds, the
+    behavior is unspecified. *)
+
+val pos : t -> int64
+(** Return the current reading position for the given channel.  For files opened
+    in text mode under Windows, the returned position is approximate (owing to
+    end-of-line conversion); in particular, saving the current position with
+    {!pos}, then going back to this position using {!seek} will not work.  For
+    this programming idiom to work reliably and portably, the file must be
+    opened in binary mode. *)
+
+val length : t -> int64
+(** Return the size (number of characters) of the regular file on which the
+    given channel is opened.  If the channel is opened on a file that is not a
+    regular file, the result is meaningless.  The returned size does not take
+    into account the end-of-line translations that can be performed when reading
+    from a channel opened in text mode. *)
+
+val close : t -> unit
+(** Close the given channel.  Input functions raise a [Sys_error] exception when
+    they are applied to a closed input channel, except {!close}, which does
+    nothing when applied to an already closed channel. *)
+
+val close_noerr : t -> unit
+(** Same as {!close}, but ignore all errors. *)
+
+val input_char : t -> char option
+(** Read one character from the given input channel.  Returns [None] if there
+    are no more characters to read. *)
+
+val input_byte : t -> int option
+(** Same as {!input_char}, but return the 8-bit integer representing the
+    character. Returns [None] if an end of file was reached. *)
+
+val input_line : t -> string option
+(** Read characters from the given input channel, until a newline character is
+    encountered. Return the string of all characters read, without the newline
+    character at the end.  Returns [None] if the end of the file is reached at
+    the beginning of line. *)
```

Comment:
Thanks, I pushed a commit with your suggestion, except that I kept the existing formulation

> Returns [None] if the end of the file is reached at the beginning of line.

instead of your suggestion

> Returns [None] if the end of file is reached or if the last line is empty

(I find the existing formulation clearer)

Reply:
I don't find the existing formulation very clear:

1. In all other doc strings the bit about [None] fully specifies when it's returned, I think it should also here, and it does not. It's not only if "the end of the file is reached at the beginning of line."
2. I won't fight about this but I find the notion of "end of file at the beginning of line", to be a bit of a mind twister: it's like of you have an end of file but there's still a line after. In my opinion, my formulation makes it more clear that a trailing empty line will be stripped.

Reply:
I also won't fight over this :) but just for my own clarity:

> 1. In all other doc strings the bit about [None] fully specifies when it's returned, I think it should also here, and it does not. It's not only if "the end of the file is reached at the beginning of line."

I thought this was indeed the case (which is why I said that I found this formulation clearer). In which other case does this function return `None`?

Reply:
When the end of file happens at the end of a line.

Reply:
Right, I see what you mean. Fair enough, docstring amended!

Reply:
But sorry, I think my `if the last line is empty` formulation is also a bit ambiguous. We should say one of these:

 * `if the last line of input is empty`  
 * `if the line before end of file is empty`

Reply:
Actually, I'm wondering if `Returns [None] if the end of file is reached` wouldn't be enough to describe all cases when the function returns `None`. Namely, this happens whenever the current file position (marked with a `p`) just before the call to the function is `xxxxp<EOF>` or `xxxx\np<EOF>`. In both cases, the end of file has been reached.

Reply:
> Actually, I'm wondering if `Returns [None] if the end of file has been reached` wouldn't be enough to describe all cases when the function returns `None`

I think you are right from a specification point of view, but the fact that it took us all this exchange to get to realise that is a sign that's it's too subtle :-)

People have been caught in the past by the fact that say `input_line` in a list until eof and then `String.concat "\n" lines` won't be able to recreate your initial input in all cases. 

So I'd rather be more explicit to avoid surprises and head scratches.

Reply:
OK, what about this version? I replaced `is reached` by `has been reached` because EOF should have already been reached when the function is called for `None` to be returned.
And I added the note about the last line of input in a separate sentence, as it is just clarifying the general rule.
```
[input_line ic] reads characters from [ic] until a newline or the end of
file is reached.  Returns the string of all characters read, without the
newline (if any).  Returns [None] if the end of the file has been reached.
In particular, this will be the case if the last line of input is empty.
```

Reply:
Reads ok to me !